datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java

```diff
@@ -27,6 +27,8 @@ public class AddTagResolver implements DataFetcher<CompletableFuture<Boolean>> {
 
   @Override
   public CompletableFuture<Boolean> get(DataFetchingEnvironment environment) throws Exception {
+    // ETAG Comment: Client should send the eTag as part of the Variables. Here we received it for GraphQL implementation.
+    final String eTag = environment.getVariables().containsKey("eTag") ? environment.getVariables().get("eTag").toString() : null;
```

Comment:
This seems to imply the ETag is expected to come from the client rather than from the server. Per the RFC, client should be sending in the If-Match header (or perhaps [If-Unmodified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since) makes more sense here) and the server should be generating the ETag to be returned as a response header:

```
We could achieve this via HTTP [Conditional Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#avoiding_the_lost_update_problem_with_optimistic_locking).
GMS would populate the `ETag` header in the GET response for aspects. This can be used in the `If-Match` or 
`If-None-Match` header when a POST request is passed to GMS to update the aspects. The main concern is then how to 
format the ASCII `Etag` value for multiple aspects such that client and server are able to compose them.

A suggested format would be semicolon-separated aspect-value pairs, e.g. `myAspect1=hashValue1;myAspect2=hashValue2`.
```

Clients don't typically send in the ETag header to my understanding as it's a server side identification concept to be used in serial requests dependent on eachother. I.E.:

Client -> Server (ETag: XYZ) -> Client (If-Match: XYZ) -> Server

Reply:
It is doable. After some internal discussion we had about this, and also about condition insert, here are some conclusions and suggestions for the next steps:
1) Instead of using eTag, we should use "If-Unmodified-Since" header. I somehow think this one is better choice than "If-Match".
2) We should use the same header for conditional insert. If "If-Unmodified-Since" is 0 than we do the insert only if the aspect doesn't exist. We could introduce another header for this but I would avoid it as it is simpler to have one variable.

To sum up, here are the possible values (and meanings) for "If-Unmodified-Since" header:
1) If "If-Unmodified-Since" is greater then 0 we do the conditional update. 
2) If "If-Unmodified-Since" is equals 0 we do the conditional insert. I see "If-Unmodified-Since = 0" as "If Unmodified Since the beginning of the time == If not Created Yet"
3) If "If-Unmodified-Since" is null we do NOT apply any conditional operation.

For GraphQL I would suggest to use variable with the same name "If-Unmodified-Since" to provide the information.

Reply:
Yep having 0 have meaning makes sense to me, this is not a default header that gets added so if someone is adding this header in it is expected to actually mean something. Using If-Unmodified-Since makes more sense to me as well since we're using a timestamp driven approach.

Reply:
@RyanHolstien, note that `If-Unmodified-Since == 0` is effectively the `CREATE` [ChangeType](https://github.com/datahub-project/datahub/blob/master/metadata-models/src/main/pegasus/com/linkedin/events/metadata/ChangeType.pdl). I'd suggest a follow-up PR later on that could wire this together if there is interest.